src/Database/LSMTree/Normal.hs

```diff
@@ -351,11 +357,19 @@ snapshot snap (TableHandle th) = void $ Internal.snapshot snap th
 -- TODO: this function currently has a temporary implementation until we have
 -- proper snapshots. See 'snapshot'.
 open ::
-     (IOLike m, SerialiseKey k, SerialiseValue v, SerialiseValue blob)
+     forall m k v blob. ( IOLike m
+     , SerialiseKey k, SerialiseValue v, SerialiseValue blob
+     , Typeable k, Typeable v, Typeable blob
+     )
   => Session m
   -> SnapshotName
   -> m (TableHandle m k v blob)
-open (Session sesh) snap = TableHandle <$> Internal.open sesh snap
+open (Session sesh) snap = do
+    (label, th) <- Internal.open sesh snap
+    unless (label == label') $ throwIO (Internal.ErrSnapshotWrongType snap)
+    pure $! TableHandle th
+  where typeReps = (typeRep (Proxy @k), typeRep (Proxy @v), typeRep (Proxy @blob))
+        label' = show typeReps
```

Comment:
On reflection, I don't think we can do this kind of check reliably using Typeable. I'm not sure that the type rep is actually guaranteed stable over GHC versions, or library versions. That'd result in the check failing when the (de)serialiser is still compatible.

We may need to punt a dynamic type sanity check to user code, e.g. ask them to produce a signature or version or something. That could be part of the type class. Or just not do it at all.

Anyway, I think don't do a typeable based one here now. We can leave dynamic checks as a TODO for later.